### ISSUE TITLE
Problem/question regarding the init procedure

### ISSUE THREAD

**Issue by dagniel** [telefonicaid/fiware-orion#3515]
Hello all,

I'm trying to deploy the Orion ContextBroker on a Openshift/OKD(Kubernetes) cluster and I'm having a problem with its deployment regarding the initialization time.

I'm using the 2.2.0 release tag for the ContextBroker with mongoDB 3.2.0. The startup args for Orion are:
```
-ipv4 -reqPoolSize 100 -notificationMode threadpool:10000:50 -statNotifQueue -statCounters -statSemWait -statTiming -relogAlarms -httpTimeout 100000
```

The initialization appears to be somewhat inconsistent regarding the time required for the app to become available.
Sometimes the deployment runs "smoothly" and sometimes the app fails to start (in a reasonable time interval).

The initialization seems to freeze at a certain point which appears in the logs as [[1]](#log_1). The actual service isn't started (`lsof -i -n -P` doesn't return any processes using port 1026).
I use standard health-checks which, basically, do a `curl localhost:1026/version`. I've tried modifying the timeouts and also the delay time from which the probe fires. Not even with 360sec (6 minutes) delay do I reach consistent deployments!

I've tested with different resource allocation and this doesn't seem to be the problem.

Also by checking the logs I see some "odd" intervals in the initialization procedure. I have some excerpts at the end of the message,  [[2]](#log_2), where I can see the last steps of the init procedure being executed(or, at least, logged) at precisely 1 minute intervals.

The problem is that once the readiness health check fails, the deployment fails as well. Orion seems to use a lot of RAM which does not get released even if the notification load disappears, from what I saw. The recommendation would be to restart the process, which, in my case, can be automatically handled if I set an upper memory limit for the container. So the initialization process comes again in question...also for the auto-scaling mechanism, etc.

Any hints towards how to solve this problem would be much appreciated!

Thanks, Dan

==========================

<a name="log_1">1</a> - last 10-12 lines from ContextBroker's log; DEBUG -t 0-255

```
time=Wednesday 19 Jun 16:50:35 2019.407Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[802]:getWriteConcern | msg=getWriteConcern()
time=Wednesday 19 Jun 16:50:35 2019.407Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[807]:getWriteConcern | msg=Database Operation Successful (getWriteConcern)
time=Wednesday 19 Jun 16:50:35 2019.407Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=mongoConnectionPool.cpp[240]:mongoConnect | msg=Active DB Write Concern mode: 1
time=Wednesday 19 Jun 16:50:35 2019.431Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[691]:runCollectionCommand | msg=runCommand() in 'admin' collection: '{ buildinfo: 1 }'
time=Wednesday 19 Jun 16:50:35 2019.431Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[701]:runCollectionCommand | msg=Database Operation Successful (command: { buildinfo: 1 })
time=Wednesday 19 Jun 16:50:35 2019.431Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=mongoConnectionPool.cpp[282]:mongoConnect | msg=mongo version server: 3.2.10 (mayor: 3, minor: 2, extra: 10)
time=Wednesday 19 Jun 16:50:35 2019.432Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=MongoGlobal.cpp[227]:mongoInit | msg=Connected to mongo at mongodb.civ-fiware.svc.cluster.local:admin as user 'admin'
time=Wednesday 19 Jun 16:50:35 2019.432Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=MongoGlobal.cpp[448]:setDbPrefix | msg=Set dbPrefix to 'admin'
time=Wednesday 19 Jun 16:50:35 2019.432Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=MongoGlobal.cpp[607]:composeDatabaseName | msg=database name composed: 'admin'
time=Wednesday 19 Jun 16:50:35 2019.435Z | lvl=DEBUG | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[577]:collectionCreateIndex | msg=createIndex() in 'admin.entities' collection: '{ location.coords: "2dsphere" }'
```
<a name="log_2">2</a> excerpt from final log entries of the initialization process; highlighted the timestamps for subsequent operations which happen to be at precisely 1 min interval; final entries are from the liveness probes when the app starts being available

<pre>
time=Wednesday 19 Jun 19:06:01 2019.060Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun <b>19:06:01 2019.107Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1040]:main | msg=Startup completed
time=Wednesday 19 Jun <b>19:07:01 2019.062Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun 19:07:01 2019.064Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastFailure: { $lt: 1560855211 } }, { lastFailure: { $exists: false } } ] }, { $set: { lastFailure: 1560855211, lastFailureReason: "Timeout was reached" } }>)
time=Wednesday 19 Jun <b>19:07:01 2019.064Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastSuccess: { $lt: 1560855687 } }, { lastSuccess: { $exists: false } } ] }, { $set: { lastSuccess: 1560855687, lastSuccessCode: 200 } }>)
time=Wednesday 19 Jun <b>19:08:01 2019.066Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun 19:08:01 2019.067Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastFailure: { $lt: 1560855211 } }, { lastFailure: { $exists: false } } ] }, { $set: { lastFailure: 1560855211, lastFailureReason: "Timeout was reached" } }>)
time=Wednesday 19 Jun <b>19:08:01 2019.067Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastSuccess: { $lt: 1560855687 } }, { lastSuccess: { $exists: false } } ] }, { $set: { lastSuccess: 1560855687, lastSuccessCode: 200 } }>)
time=Wednesday 19 Jun <b>19:09:01 2019.069Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun 19:09:01 2019.071Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastFailure: { $lt: 1560855211 } }, { lastFailure: { $exists: false } } ] }, { $set: { lastFailure: 1560855211, lastFailureReason: "Timeout was reached" } }>)
time=Wednesday 19 Jun <b>19:09:01 2019.071Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastSuccess: { $lt: 1560855687 } }, { lastSuccess: { $exists: false } } ] }, { $set: { lastSuccess: 1560855687, lastSuccessCode: 200 } }>)
time=Wednesday 19 Jun <b>19:10:01 2019.073Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun 19:10:01 2019.075Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastFailure: { $lt: 1560855211 } }, { lastFailure: { $exists: false } } ] }, { $set: { lastFailure: 1560855211, lastFailureReason: "Timeout was reached" } }>)
time=Wednesday 19 Jun <b>19:10:01 2019.076Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastSuccess: { $lt: 1560855687 } }, { lastSuccess: { $exists: false } } ] }, { $set: { lastSuccess: 1560855687, lastSuccessCode: 200 } }>)
time=Wednesday 19 Jun <b>19:11:01 2019.078Z</b> | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Wednesday 19 Jun 19:11:01 2019.079Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastFailure: { $lt: 1560855211 } }, { lastFailure: { $exists: false } } ] }, { $set: { lastFailure: 1560855211, lastFailureReason: "Timeout was reached" } }>)
time=Wednesday 19 Jun 19:11:01 2019.080Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[449]:collectionUpdate | msg=Database Operation Successful (update: <{ _id: ObjectId('5d0215c1294122a2ec462ff3'), $or: [ { lastSuccess: { $lt: 1560855687 } }, { lastSuccess: { $exists: false } } ] }, { $set: { lastSuccess: 1560855687, lastSuccessCode: 200 } }>)
time=Wednesday 19 Jun 19:11:10 2019.809Z | lvl=INFO | corr=ffb9d5e6-92c5-11e9-885f-0a580a8301d0 | trans=1560971160-541-00000000001 | from=10.131.0.1 | srv=<none> | subsrv=<none> | comp=Orion | op=logMsg.h[1844]:lmTransactionStart | msg=Starting transaction from 10.131.0.1:58238/version
time=Wednesday 19 Jun 19:11:10 2019.820Z | lvl=INFO | corr=ffb9d5e6-92c5-11e9-885f-0a580a8301d0 | trans=1560971160-541-00000000001 | from=10.131.0.1 | srv=<none> | subsrv=<none> | comp=Orion | op=logMsg.h[1874]:lmTransactionEnd | msg=Transaction ended
time=Wednesday 19 Jun 19:11:25 2019.800Z | lvl=INFO | corr=08a93c1e-92c6-11e9-b713-0a580a8301d0 | trans=1560971160-541-00000000002 | from=10.131.0.1 | srv=<none> | subsrv=<none> | comp=Orion | op=logMsg.h[1844]:lmTransactionStart | msg=Starting transaction from 10.131.0.1:57998/version
time=Wednesday 19 Jun 19:11:25 2019.801Z | lvl=INFO | corr=08a93c1e-92c6-11e9-b713-0a580a8301d0 | trans=1560971160-541-00000000002 | from=10.131.0.1 | srv=<none> | subsrv=<none> | comp=Orion | op=logMsg.h[1874]:lmTransactionEnd | msg=Transaction ended
</pre>

**Comment by fgalan** [comment]
I'm afraid I cannot provide too much feedback, as I'm not an expert in Openshift/OKD(Kubernetes). Have you tested the same setup in a different platform (e.g. plain docker containers using docker-compose.yml for instance)? Just to check how it behaves there...

Some additional pieces of feedback:

* MongoDB 3.2 is two steps behind the version recommended for Orion, which is MongoDB 3.6 (see https://fiware-orion.readthedocs.io/en/master/admin/install/index.html#requirements). Note that 3.2 reached end of life in september last year (https://www.mongodb.com/support-policy).
* The traces you see at 1 minute interval are normal. It is due to subscription cache refresh logic which by default takes place every 60 seconds (this can be changed with `-subCacheIval`)
* Orion uses dynamic memory as it needs and also frees it. However, note that not always the release of memory in the program is immediately returned to the operating system (for instance due to memory reuse in the next allocation). This is normal in general for all kind of processes (you can have a look to some discussiong about it [here](https://stackoverflow.com/questions/5365996/memory-not-freed-after-calling-free) and [here](https://stackoverflow.com/questions/15139436/why-free-doesnt-really-frees-memory)

**Comment by dagniel** [comment]
Thanks for the reply.

* I've missed the EOL detail for MongoDB 3.2. Now I have switched to MongoDB 3.6.
* Regarding the release of memory, I wasn't expecting immediate release, but what I've seen is that memory usage grew as high as 2.8GB and it didn't drop not even hours after the notification load has been stopped completely.

I haven't actually tested with docker-compose, however, the supporting framework where I'd want this done is Openshift.
The underlying architecture in Openshift is similar to the one from Docker compose; they are actual Docker containers but just orchestrated  by a higher level process.
The only difference that I can think of, regarding the actual Orion container, is that I use a wrap-up script for starting the process as a convenient way to configure start-up args. This makes it that the main contextBroker process is actually PID=6 instead of PID=1. If this would somehow make a difference in behaviour....Also, as the deployment is managed, this means that, if the container doesn't get to "ready" status, it is restarted a couple of times until it is marked as failed. Restarting involves redoing the init procedure all over again.

Getting back to my problem, I've tested with the new version of Mongo and this doesn't change much.
I have failed deployments with same behaviour as noted initially. The final log entry for all of them is this:
```
time=Monday 24 Jun 09:32:17 2019.632Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=MongoGlobal.cpp[227]:mongoInit | msg=Connected to mongo at mongodb.civ-fiware.svc.cluster.local:admin as user 'admin'
```

Somehow subsequent tries/deployments succeed immediately and the logs are like this:
```
time=Monday 24 Jun 09:32:17 2019.632Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=MongoGlobal.cpp[227]:mongoInit | msg=Connected to mongo at mongodb.civ-fiware.svc.cluster.local:admin as user 'admin'
time=Monday 24 Jun 09:32:18 2019.023Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[597]:collectionCreateIndex | msg=Database Operation Successful (createIndex: { location.coords: "2dsphere" })
time=Monday 24 Jun 09:32:18 2019.095Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[597]:collectionCreateIndex | msg=Database Operation Successful (createIndex: { expDate: 1 })
time=Monday 24 Jun 09:32:18 2019.116Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=connectionOperations.cpp[94]:collectionQuery | msg=Database Operation Successful (query: {})
time=Monday 24 Jun 09:32:18 2019.135Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1040]:main | msg=Startup completed
```

It appears, judging from the middle 3 lines of log, that there are some DB operations being done. Can it be that the initial deployment(s) start these operations and they are completed only at a later time so the subsequent deployment of Orion gets to "ready" state right away?

Thank you,
Dan

**Comment by fgalan** [comment]
Orion connects to DB at startup and immediately does some DB initialization operations (the ones you see in the log about index creation). It doesn't wait to complete them in a later time. This can be easily seen in the code of the [mongoInit() function](https://github.com/telefonicaid/fiware-orion/blob/e3f31930f2d9532747b179303ba698c604206615/src/lib/mongoBackend/MongoGlobal.cpp#L203) doing this connection and initialization. 

In your case it is like in some cases Orion fails just after connection to the database... but some other times it works fine. Maybe some kind of instability in the Orion-to-MongoDB connection in the underlying system (docker, OS, Openshift, etc.)? Maybe MongoDB logs could provide some useful information about what's going on? Maybe OpenShift or docker container logs?

Some additional comment:

1. Is the result the same if you run Orion without the wrapping script so it gets PID 1 in the docker container? (just to check what you say "If this would somehow make a difference in behavior"... it shoudn't, but, who knows...)
2. Even if your target platform is Openshift, maybe it works the time to test with vanilla docker-compose. If it works, then the problems is potentially in the Orion and Openshift combination. If it fails, the problem is not related with OpenShift and it could be reproducible in conventional hosts (I don't have any OpenShift instance at hand ;). No matter one case of the other, the result is a piece of knowledge that could help in the debugging work of this issue.

**Comment by dagniel** [comment]
Hello and thanks for the support!

I found the issue which was causing this behaviour and it is a piece of configuration that I overlooked.

More to the point: what I haven't mentioned in my initial description(somehow I didn't think it would be a possible cause) is that I have set up a **replica set** for MongoDB.
I have tested with a standalone MongoDB with which the contextBroker initializes perfectly fine, but when switching to the replicaSet configuration for Mongo the behaviour initially described, appears.

So I have revisited the [Context Broker documentation about start up arguments](https://fiware-orion.readthedocs.io/en/master/admin/cli/index.html#command-line-options) and found the **-rplSet** start up argument, which, apparently is pretty important..:) When adding this as a start up argument, the Pod/container for the context broker initializes in a consistent manner.

So, to sum up, the actual issue in my case was, in fact, that the Context Broker initialized at all, given the fact that it uses the standalone method to connect and not the replica Set one and the connection is successful on first try, by looking at the code [here](https://github.com/telefonicaid/fiware-orion/blob/e3f31930f2d9532747b179303ba698c604206615/src/lib/mongoBackend/mongoConnectionPool.cpp#L131)
